### DISCUSSION TITLE
How to get server session to lock when connected via xrdp

### DISCUSSION THREAD

**Discussion by BusterRaid** [neutrinolabs/xrdp#2963] [Q&A]
Hi.
I cant seem to get the session to “lock” when connected via xrdp, either on-demand or via “_Power management/Blank after X mintues_”.
The device running xrdp server is a VM running  Debian 12.5 Bookworm with xfce4 desktop and the client is Windows 10  Pro, connecting via "built it" Remote Desktop Client.

### **Connected “directly”, via VMWare Player:**
When connected “directly”, via VMWare Player, I can lock the screen with:

`light-locker-command -l`

Or:

`dm-tool lock`

Or:

… if the display is set to “_Blank after X minutes_” in “Power Management”, the screen will automagically lock, after the allotted time.

The screen briefly goes black and then the greeter is presented.

... and examiningthe log (journalctl), there are many entries regarding closing and stopping stuff.

### **Connected via xrdp:**
When connected via xrdp, issuing the following command in a Terminal (as the logged in user):

`light-locker-command -l`

… returns:

> ** Message: 17:45:28.147: Received error message from the locker: GDBus.Error:org.freedesktop.DBus.Error.ServiceUnknown: The name org.freedesktop.ScreenSaver was not provided by any .service files

... and examining the log (journalctl), there are just two entries:

> Feb 22 18:35:36 XRDEB12VM xrdp-chansrv[5020]: [ERROR] clipboard_event_selection_request: unknown target text/plain;charset=utf-8
> Feb 22 18:35:36 XRDEB12VM xrdp-chansrv[5020]: [ERROR] clipboard_event_selection_request: unknown target text/plain;charset=utf-8

... and looking at _xrdp.log_ ,_xrdp-sesman.log_ and _/var/log/lightdm/seat0-greeter.log_, there are no entries created when the above command is issued.

When issuing the following command in a Terminal (as the logged in user):

`dm-tool lock`

… returns:

> Not running inside a display manager, XDG_SEAT_PATH not defined

Looking at “light-locker” (https://github.com/the-cavalry/light-locker/issues?q=rdp), I couldnt see any items regarding rdp.

Any help or ideas would be appreciated.
Thank you.
 
 
**Client:** 
Windows 10 Pro x64 2H22
Remote Desktop Connection client.
Remote Desktop Connection Shell\Control Version: 10.0.19041
Remote Desktop Protocol 10.8

**Server:**
VM running under VMWare Player.
OS: Debian 12.5 Bookworm x86-64
Kernel: Linux 6.1.0-18-amd64

**Software:**
xorgxrdp/stable,now 1:0.9.19-1 amd64 
xrdp/stable,now 0.9.21.1-1 amd64 
light-locker/stable,now 1.8.0-3 amd64
xfce4/stable,now 4.18 all 
 
Run via xrdp - `systemctl status display-manager:`

> ● lightdm.service - Light Display Manager
>      Loaded: loaded (/lib/systemd/system/lightdm.service; enabled; preset: enabled)
>      Active: active (running) since Thu 2024-02-22 09:53:49 GMT; 7h ago
>        Docs: man:lightdm(1)
>    Main PID: 677 (lightdm)
>       Tasks: 6 (limit: 4575)
>      Memory: 98.6M
>         CPU: 3min 11.487s
>      CGroup: /system.slice/lightdm.service
>              ├─  677 /usr/sbin/lightdm
>              ├─11363 /usr/lib/xorg/Xorg :0 -seat seat0 -auth /var/run/lightdm/root/:0 -nolisten tcp vt7 -novtswitch
>              └─11441 lightdm --session-child 14 22

Run via VMWare - `systemctl status display-manager:`

> ● lightdm.service - Light Display Manager
>      Loaded: loaded (/lib/systemd/system/lightdm.service; enabled; preset: enabled)
>      Active: active (running) since Thu 2024-02-22 17:50:11 GMT; 29min ago
>        Docs: man:lightdm(1)
>    Main PID: 682 (lightdm)
>       Tasks: 5 (limit: 4575)
>      Memory: 150.1M
>         CPU: 18.423s
>      CGroup: /system.slice/lightdm.service
>              ├─ 682 /usr/sbin/lightdm
>              └─3541 /usr/lib/xorg/Xorg :0 -seat seat0 -auth /var/run/lightdm/root/:0 -nolisten tcp vt7 -novtswitch

Run via xrdp - `ME@XRDEB12VM:~$ env:`

> SESSION_MANAGER=local/XRDEB12VM:@/tmp/.ICE-unix/5016,unix/XRDEB12VM:/tmp/.ICE-unix/5016
> XDG_CONFIG_DIRS=/etc/xdg
> XDG_MENU_PREFIX=xfce-
> DESKTOP_SESSION=xfce
> GTK_MODULES=gail:atk-bridge
> XDG_SESSION_TYPE=x11
> PANEL_GDK_CORE_DEVICE_EVENTS=0
> PULSE_SCRIPT=/etc/xrdp/pulse/default.pa
> XDG_CURRENT_DESKTOP=XFCE
> VTE_VERSION=7006
> XRDP_SOCKET_PATH=/run/xrdp/sockdir
> XRDP_PULSE_SOURCE_SOCKET=xrdp_chansrv_audio_in_socket_10
> XDG_SESSION_CLASS=user
> TERM=xterm-256colo
> XRDP_SESSION=1
> DISPLAY=:10.0
> XRDP_PULSE_SINK_SOCKET=xrdp_chansrv_audio_out_socket_10
> SHLVL=1
> XDG_SESSION_ID=c10
> XDG_RUNTIME_DIR=/run/user/1000
> XDG_DATA_DIRS=/usr/share/xfce4:/usr/local/share/:/usr/share/:/usr/share
> PATH=/usr/local/bin:/usr/bin:/bin:/usr/local/games:/usr/games
> DBUS_SESSION_BUS_ADDRESS=unix:path=/run/user/1000/bus
> UID=1000
> _=/usr/bin/env

Run via VMWare- `ME@XRDEB12VM:~$ env:`

> SESSION_MANAGER=local/XRDEB12VM:@/tmp/.ICE-unix/4305,unix/XRDEB12VM:/tmp/.ICE-unix/4305
> XDG_CONFIG_DIRS=/etc/xdg
> XDG_SESSION_PATH=/org/freedesktop/DisplayManager/Session2
> XDG_MENU_PREFIX=xfce-
> DESKTOP_SESSION=xfce
> GTK_MODULES=gail:atk-bridge
> XDG_SEAT=seat0
> XDG_SESSION_DESKTOP=xfce
> XDG_SESSION_TYPE=x11
> PANEL_GDK_CORE_DEVICE_EVENTS=0
> XAUTHORITY=/home/me/.Xauthority
> XDG_GREETER_DATA_DIR=/var/lib/lightdm/data/me
> XDG_CURRENT_DESKTOP=XFCE
> VTE_VERSION=7006
> XDG_SEAT_PATH=/org/freedesktop/DisplayManager/Seat0
> XDG_SESSION_CLASS=user
> TERM=xterm-256color
> DISPLAY=:0.0
> SHLVL=1
> XDG_VTNR=7
> XDG_SESSION_ID=32
> XDG_RUNTIME_DIR=/run/user/1000
> XDG_DATA_DIRS=/usr/share/xfce4:/usr/local/share/:/usr/share/:/usr/share
> PATH=/usr/local/bin:/usr/bin:/bin:/usr/local/games:/usr/games
> GDMSESSION=xfce
> DBUS_SESSION_BUS_ADDRESS=unix:path=/run/user/1000/bus
> _=/usr/bin/env
> 
> 
> 
>

**Comment by matt335672** [comment]
- Log in on the VMWare player console.
- Find out the PID of the executable that provides `org.freedesktop.ScreenSaver` with this command:-

   ```
   dbus-send --session --print-reply --dest=org.freedesktop.DBus / org.freedesktop.DBus.GetConnectionUnixProcessID string:org.freedesktop.ScreenSaver
   ```

Then you can use `ps -fp` on that PID to find out what the executable is.

Here's an example from my Linux Mint machine running XFCE:-

<pre>
$ dbus-send --session --print-reply --dest=org.freedesktop.DBus / org.freedesktop.DBus.GetConnectionUnixProcessID string:org.freedesktop.ScreenSaver
method return time=1708703155.300133 sender=org.freedesktop.DBus -> destination=:1.5157 serial=3 reply_serial=2
   uint32 <b>3049</b>
$ ps -fp <b>3049</b>
UID          PID    PPID  C STIME TTY          TIME CMD
...         3049    2951  0 10:17 ?        00:00:00 xscreensaver-systemd -verbos
</pre>

We can then figure out if that executable should be running in your XRDP session, and if it should, why it isn't.

**Comment by BusterRaid** [comment]
Thank you for the reply.

Via VMWare Player console:

`$ dbus-send --session --print-reply --dest=org.freedesktop.DBus / org.freedesktop.DBus.GetConnectionUnixProcessID string:org.freedesktop.ScreenSaver`

... returns: 

> method return time=1708707794.676135 sender=org.freedesktop.DBus -> destination=:1.830 serial=3 reply_serial=2
>    uint32 8428

Looking up that PID returns:

> UID          PID    PPID  C STIME TTY          TIME CMD
> xxxxx+    8428    8264  0 16:41 ?        00:00:00 light-locker

**Comment by matt335672** [comment]
Try running light-locker by hand in your session and see what happens.

**Comment by BusterRaid** [comment]
All via xrdp connection, "normal" user:

`$ light-locker`

... Returns:

> ** (light-locker:12656): ERROR **: 18:33:18.210: Environment variable XDG_SESSION_PATH not set. Is LightDM running?
> Trace/breakpoint trap

The absence of "_XDG_SESSION_PATH_" is as in the OP.
For VMWare, XDG_SESSION_PATH is set.
Via xrdp, it is **not** present.

Running:

`$ /usr/sbin/lightdm`

... Returned:

> Only root can run Light Display Manager.  To run as a regular user for testing run with the --test-mode flag.

... Running:

`$ /usr/sbin/lightdm --test-mode`

Returned:

> Running inside an X server requires Xephyr to be installed but it cannot be found.  Please install it or update your PATH environment variable.

I then ran (still via xrdp, "normal" user):

`$ ps -ef | grep lightdm`

... And it rather surprisingly returned:

> 
> root         682       1  0 16:48 ?        00:00:00 /usr/sbin/lightdm
> root       12022     682  0 18:32 tty7     00:00:00 /usr/lib/xorg/Xorg :0 -seat seat0 -auth /var/run/lightdm/root/:0 -nolisten tcp vt7 -novtswitch
> root       12029     682  0 18:32 ?        00:00:00 lightdm --session-child 19 22
> lightdm    12033       1  0 18:32 ?        00:00:00 /lib/systemd/systemd --user
> lightdm    12034   12033  0 18:32 ?        00:00:00 (sd-pam)
> lightdm    12049   12033  0 18:32 ?        00:00:00 /usr/bin/pulseaudio --daemonize=no --log-target=journal
> lightdm    12050   12029  0 18:32 ?        00:00:01 /usr/sbin/lightdm-gtk-greeter
> lightdm    12055   12033  0 18:32 ?        00:00:00 /usr/bin/dbus-daemon --session --address=systemd: --nofork --nopidfile --systemd-activation --syslog-only
> lightdm    12059   12033  0 18:32 ?        00:00:00 /usr/libexec/at-spi-bus-launcher
> lightdm    12066   12059  0 18:32 ?        00:00:00 /usr/bin/dbus-daemon --config-file=/usr/share/defaults/at-spi2/accessibility.conf --nofork --print-address 11 --address=unix:path=/run/user/107/at-spi/bus
> lightdm    12068   12033  0 18:32 ?        00:00:00 /usr/libexec/gvfsd
> lightdm    12073   12033  0 18:32 ?        00:00:00 /usr/libexec/gvfsd-fuse /run/user/107/gvfs -f
> root       12099     682  0 18:32 ?        00:00:00 lightdm --session-child 14 22
> lightdm    12101   12033  0 18:32 ?        00:00:00 /usr/libexec/at-spi2-registryd --use-gnome-session
> chrisbe+   12695   12650  0 18:37 pts/0    00:00:00 grep lightdm
> 

... Suggestive to my not so clear eye, that lightdm is running.
Checking the lightdm.service, the following is seen:

> lightdm.service - Light Display Manager
>      Loaded: loaded (/lib/systemd/system/lightdm.service; enabled; preset: enabled)
>      Active: active (running) since Thu 2024-02-22 17:50:11 GMT; 1 day 1h ago
>        Docs: man:lightdm(1)
>    Main PID: 682 (lightdm)
>       Tasks: 6 (limit: 4575)
>      Memory: 95.5M
>         CPU: 54.737s
>      CGroup: /system.slice/lightdm.service
>              ├─  682 /usr/sbin/lightdm
>              ├─13999 /usr/lib/xorg/Xorg :0 -seat seat0 -auth /var/run/lightdm/root/:0 -nolisten tcp vt7 -novtswitch
>              └─14081 lightdm --session-child 15 22
> 
> Feb 23 18:32:45 XRDEB12VM lightdm[12029]: Error getting user list from org.freedesktop.Accounts: GDBus.Error:org.freedesktop.>
> Feb 23 18:32:45 XRDEB12VM lightdm[12029]: pam_unix(lightdm-greeter:session): session opened for user lightdm(uid=107) by (uid>
> Feb 23 18:39:03 XRDEB12VM lightdm[13395]: gkr-pam: unable to locate daemon control file
> Feb 23 18:39:03 XRDEB12VM lightdm[13395]: gkr-pam: stashed password to try later in open session
> Feb 23 18:39:03 XRDEB12VM lightdm[13395]: Error getting user list from org.freedesktop.Accounts: GDBus.Error:org.freedesktop.>
> Feb 23 18:39:03 XRDEB12VM lightdm[682]: g_dbus_connection_call_sync_internal: assertion 'G_IS_DBUS_CONNECTION (connection)' f>
> Feb 23 18:39:03 XRDEB12VM lightdm[13395]: pam_unix(lightdm:session): session opened for user chrisbennett(uid=1000) by (uid=0)
> Feb 23 18:39:03 XRDEB12VM lightdm[13395]: gkr-pam: unlocked login keyring
> Feb 23 18:49:19 XRDEB12VM lightdm[14007]: Error getting user list from org.freedesktop.Accounts: GDBus.Error:org.freedesktop.>
> Feb 23 18:49:19 XRDEB12VM lightdm[14007]: pam_unix(lightdm-greeter:session): session opened for user lightdm(uid=107) by (uid>
> ~

**Comment by matt335672** [comment]
@BusterRaid - thanks for that investigative work. There's some useful information there.

My guess is that you've got `lightdm` running to permit logins to the console (or 'via VMWare player' as you describe it above). When the user logs in at the console, `lightdm` makes a note of the user session so that it can control not only the screen locking but user switching.

My guess is that it sets `XDG_SESSION_PATH` appropriately so that `light-locker` can communicate with it over D-Bus.

Because `lightdm` doesn't know anything about xrdp sessions, I don't think there will be a simple way to make the `lightdm` lock screen work with xrdp.

From the little research I've done, it looks like `xscreensaver` might be able to replace this though. Can you try installing and running xscreensave in an xrdp session and see if it works for you? If so, we'll figure out a way to starting when you log in over xrdp.

**Comment by BusterRaid** [comment]
Hi.
As per suggestion, I uninstalled light-locker:

`apt remove light-locker`

... Installed xscreensaver:

```
apt update
apt install xscreensaver
```

Rebooted.

Screen lock (and screen saver) works on VMWare Play console **and** via xrdp.

Happy for this to be closed or I can help out with more testing regarding "light-locker".

Thank you.

**Reply by matt335672** [reply]
Thanks @BusterRaid - that's really useful information.

I'll close this then, as I'm pretty sure we can't get lightdm to recognise an xrdp session.
